src/github_runner_image_builder/openstack_builder.py

```diff
@@ -293,6 +306,73 @@ def run(
     return ",".join(str(image.id) for image in images)
 
 
+def _ensure_openstack_resources(
+    conn: openstack.connection.Connection, builder_name: str, key_name: str, prefix: str
+) -> None:
+    """Ensure that OpenStack resources are in expected state.
+
+    1. Ensure the key that is installed matches what is expected by OpenStack.
+    2. Ensure that only a single security group exists.
+    3. Ensure that no VMs exist.
+
+    Args:
+        conn: The OpenStack connection instance.
+        builder_name: The OpenStack builder VM name that is used to build the image.
+        key_name: The OpenStack key name used to connect to the builder VM.
+        prefix: The OpenStack resource prefix.
+    """
+    # OpenStack library does not provide good type hinting
+    key: openstack.compute.v2.keypair.Keypair | None = conn.get_keypair(
+        name_or_id=key_name
+    )  # type: ignore
+    # Check fingerprint since the key may have diverged due to unforeseen circumstances.
+    if not key or key.fingerprint != _get_key_fingerprint():
+        _create_keypair(conn=conn, prefix=prefix)
+
+    security_groups: list[openstack.network.v2.security_group.SecurityGroup] = (
+        conn.search_security_groups(name_or_id=SHARED_SECURITY_GROUP_NAME)
+    )
+    if len(security_groups) != 1:
+        for security_group in security_groups:
+            conn.delete_security_group(name_or_id=security_group.id)
+        _create_security_group(conn=conn)
+
+    servers: list[openstack.compute.v2.server.Server] = conn.search_servers(
+        name_or_id=builder_name
+    )
+    if len(servers) > 0:
+        for server in servers:
+            conn.delete_server(name_or_id=server.id)
+
+
+def _get_key_fingerprint() -> str:
+    """Get the MD5 fingerprint of the ssh key.
+
+    1. Read the private PEM file.
+    2. Get the public key from the private key.
+    3. Extract the base64 part of the public key.
+    4. Generate MD5 hash.
+
+    Returns:
+        The MD5 fingerprint hash of the ssh public key.
+    """
+    with open(BUILDER_KEY_PATH, "rb") as key_file:
+        key_data = key_file.read()
+    private_key = serialization.load_pem_private_key(
+        key_data, password=None, backend=default_backend()
+    )
+    public_key = private_key.public_key()
+    public_bytes = public_key.public_bytes(
+        encoding=serialization.Encoding.OpenSSH, format=serialization.PublicFormat.OpenSSH
+    )
+    key_base64 = public_bytes.split()[1]
+    key_data = base64.b64decode(key_base64)
+    # ignore B324:hashlib use of weak MD5 - OpenStack generates keys with this algo.
+    md5_hash = hashlib.md5(key_data).hexdigest()  # nosec: B324
+    # Ignore flake8 whitespace before : error since it conflicts with black formatter.
```

Comment:
That should go into the tox.ini file since it is a global ignore

Reply:
Updated, thank you